File: web/src/lib/cnpj.ts
// CNPJ **alfanumérico** (Receita Federal / Serpro, vigente a partir de jul/2026): 12 posições
// alfanuméricas + 2 dígitos verificadores numéricos, validados por módulo 11 com o valor de
// cada caractere = `código ASCII − 48` (`'0'..'9' → 0..9`, `'A' → 17`, …). Espelho client de
// `Api.Cnpj`: a autoridade é a API — isto só formata o input e dá feedback antes do round-trip.

const WEIGHTS_FIRST = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
const WEIGHTS_SECOND = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];

// Forma canônica de armazenamento: só [0-9A-Z], em maiúsculas, sem máscara.
export function normalizeCnpj(value: string): string {
	return value.toUpperCase().replace(/[^0-9A-Z]/g, '');
}

// Insere a máscara `XX.XXX.XXX/XXXX-XX` progressivamente, formatando também entradas parciais
// enquanto se digita. Trunca em 14 posições.
export function maskCnpj(value: string): string {
	const s = normalizeCnpj(value).slice(0, 14);
	let out = s.slice(0, 2);
	if (s.length > 2) out += '.' + s.slice(2, 5);
	if (s.length > 5) out += '.' + s.slice(5, 8);
	if (s.length > 8) out += '/' + s.slice(8, 12);
	if (s.length > 12) out += '-' + s.slice(12, 14);
	return out;
}

// Verdadeiro se, normalizado, tem 14 posições e os dois DV conferem. Rejeita `00000000000000`.
export function isValidCnpj(value: string): boolean {
	const s = normalizeCnpj(value);
	if (s.length !== 14 || s === '00000000000000') return false;
	const d1 = checkDigit(s.slice(0, 12), WEIGHTS_FIRST);
	const d2 = checkDigit(s.slice(0, 12) + d1, WEIGHTS_SECOND);
	return s[12] === d1 && s[13] === d2;
}

function checkDigit(chars: string, weights: number[]): string {
	let sum = 0;
	for (let i = 0; i < chars.length; i++) sum += (chars.charCodeAt(i) - 48) * weights[i];
	const remainder = sum % 11;
	return String(remainder < 2 ? 0 : 11 - remainder);
}

File: api/lib/api/cnpj.ex
defmodule Api.Cnpj do
  @moduledoc """
  CNPJ **alfanumérico** (Receita Federal / Serpro, vigente a partir de jul/2026): 12 posições
  alfanuméricas (`0-9`, `A-Z`) + 2 dígitos verificadores **numéricos**, validados por módulo 11.

  A única mudança frente ao CNPJ clássico é o valor de cada caractere no cálculo do DV: em vez
  do dígito, usa-se `código ASCII − 48` (então `'0'..'9' → 0..9`, `'A' → 17`, …, `'Z' → 42`).
  Como os dígitos mantêm o próprio valor, a rotina é retrocompatível: um CNPJ puramente numérico
  valida pela mesma conta. Ref: Serpro, "cálculo dos DV do CNPJ alfanumérico".
  """

  # Pesos do módulo 11 (2..9 da direita p/ a esquerda, reiniciando) — mesmos do CNPJ clássico.
  @weights_first [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
  @weights_second [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]

  @doc """
  Reduz à forma canônica de armazenamento: só `[0-9A-Z]`, em maiúsculas, sem máscara.
  `nil` ou uma string sem nenhum caractere válido viram `nil`.
  """
  @spec normalize(String.t() | nil) :: String.t() | nil
  def normalize(nil), do: nil

  def normalize(value) when is_binary(value) do
    normalized = value |> String.upcase() |> String.replace(~r/[^0-9A-Z]/, "")
    if normalized == "", do: nil, else: normalized
  end

  @doc """
  Verdadeiro se, após normalizar, o valor tem 14 posições (12 alfanuméricas + 2 DV) e os dois
  dígitos verificadores conferem. Rejeita o CNPJ "vazio" `00000000000000`.
  """
  @spec valid?(String.t()) :: boolean()
  def valid?(value) when is_binary(value) do
    case normalize(value) do
      <<base::binary-size(12), d1::binary-size(1), d2::binary-size(1)>> = cnpj ->
        cnpj != "00000000000000" and
          check_digit(base, @weights_first) == d1 and
          check_digit(base <> d1, @weights_second) == d2

      _ ->
        false
    end
  end

  def valid?(_), do: false

  # Um DV: Σ (valor_do_char × peso) mod 11; resto < 2 → 0, senão 11 − resto. Sempre numérico.
  defp check_digit(chars, weights) do
    sum =
      chars
      |> String.to_charlist()
      |> Enum.zip(weights)
      |> Enum.reduce(0, fn {code, weight}, acc -> acc + (code - 48) * weight end)

    remainder = rem(sum, 11)
    digit = if remainder < 2, do: 0, else: 11 - remainder
    Integer.to_string(digit)
  end
end
